File: ios/Sources/CapacitorShakePlugin/CapacitorShakePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorShakePlugin)
public class CapacitorShakePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorShakePlugin"
    public let jsName = "CapacitorShake"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerUser", returnType: CAPPluginReturnPromise),
    ]
    
    private let implementation = CapacitorShake()
    
    public let errorApiKeyMissing = "ApiKey must be provided."
    public let errorUserIdMissing = "User Id must be provided."


    @objc func start(_ call: CAPPluginCall) {
         guard let apiKey = call.getString("apiKey") else {
            call.reject(errorApiKeyMissing)
            return
        }
        implementation.start(apiKey)
        call.resolve()
    }

    @objc func show(_ call: CAPPluginCall) {
        implementation.show()
        call.resolve()
    }

    @objc func registerUser(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject(errorUserIdMissing)
            return
        }
        implementation.registerUser(userId)
        call.resolve()
    }
}

File: android/src/main/java/com/shakebugs/CapacitorShakePlugin.java
package com.shakebugs;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorShake")
public class CapacitorShakePlugin extends Plugin {

    private final CapacitorShake implementation = new CapacitorShake();

    public static final String ERROR_APIKEY_MISSING = "ApiKey must be provided.";
    public static final String ERROR_USER_ID_MISSING = "User Id must be provided.";

    @PluginMethod
    public void start(PluginCall call) {
        String apiKey = call.getString("apiKey");

        if(apiKey == null)
        {
            throw new RuntimeException(ERROR_APIKEY_MISSING);
        }

        implementation.start(this.getActivity(), apiKey);
        call.resolve();
    }

    @PluginMethod
    public void show(PluginCall call) {
        implementation.show();
        call.resolve();
    }

    @PluginMethod
    public void registerUser(PluginCall call) {
        String userId = call.getString("userId");

        if(userId == null)
        {
            throw new RuntimeException(ERROR_USER_ID_MISSING);
        }

        implementation.registerUser(userId);
        call.resolve();
    }
}
